ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/strategy/transferstrategy/AbstractTransferStrategy.java

```diff
@@ -167,24 +183,44 @@ private synchronized void maybeSetCachedException(final Throwable t) {
 
     private void transferJobParts(final Iterable<JobPart> jobParts,
                                   final CountDownLatch countDownLatch,
-                                  final AtomicInteger numBlobsTransferred) throws IOException {
+                                  final AtomicInteger numBlobsTransferred,
+                                  final Predicate transferPredicate) throws IOException
+    {
+        final AtomicInteger numJobPartsToTransfer = new AtomicInteger(Iterables.size(jobParts));
+
         for (final JobPart jobPart : jobParts) {
-            executorService.execute(() -> {
-                try {
+            if (executorService.isShutdown()) {
+                countDownLatch.countDown();
+            } else {
+                maybeSubmitJobPartTransfer(countDownLatch, numBlobsTransferred, transferPredicate, numJobPartsToTransfer, jobPart);
+            }
+        }
+    }
+
+    private void maybeSubmitJobPartTransfer(final CountDownLatch countDownLatch, final AtomicInteger numBlobsTransferred, final Predicate transferPredicate, final AtomicInteger numJobPartsToTransfer, final JobPart jobPart) {
+        executorService.submit((Callable<Void>) () -> {
+            try {
+                if ( ! transferPredicate.test()) {
+                    countDownLatch.countDown();
+                } else {
                     transferMethod.transferJobPart(jobPart);
-                } catch (final RuntimeException e) {
-                    emitFailureAndSetCachedException(e);
-                    throw e;
-                } catch (final Exception e) {
-                    emitFailureAndSetCachedException(e);
-                    throw new RuntimeException(e);
-                } finally {
-                    jobState.blobTransferredOrFailed(jobPart.getBlob());
-                    numBlobsTransferred.decrementAndGet();
+                }
+            } catch (final RuntimeException e) {
+                emitFailureAndSetCachedException(e);
+                throw e;
+            } catch (final Exception e) {
+                emitFailureAndSetCachedException(e);
+                throw new RuntimeException(e);
+            } finally {
+                jobState.blobTransferredOrFailed(jobPart.getBlob());
+                numBlobsTransferred.decrementAndGet();
+                if (numJobPartsToTransfer.decrementAndGet() <= 0) {
```

Comment:
If we go negative, that seems like a logic bug and is something we should log, instead of just preventing against.

Reply:
I think this line we don't know where the catch  happened, so it could already be 0?